---
layout: page
title: Genome-wide association studies
---

# Genome-wide association studies

### Computing Minor Allele Frequency (MAP) for each SNP

The allele frequency for diploids is given by

$$ f(A) = \frac{2 \times f(AA) + f(AB)}{2\times populationsize} $$
$$ f(B) = \frac{2 \times f(BB) + f(AB)}{2\times populationsize} $$

First, we load all required libraries and data files.
```{r}
library(dplyr) #  for manipulating tabular data
dat <- read.csv("genotype.csv")
```

Though we already know that we have genotype information for 9088 SNPs from 279 individuals, we avoid hard-coding our population size, and instead calculate it from the dimensions of the tabular data we have.

```{r}
population_size <- dim(dat)[1]
```

Now we can compute the Minor Allele Frequency (MAF) for each SNP.

```{r}
for (i in 2:dim(dat)[2]) { # for each column
	col <- select(dat, i)
	
	AA <- filter(col, col==0)
	AB <- filter(col, col==1)
	BB <- filter(col, col==2)
	
	f_A <- (2 * dim(AA)[1] + dim(AB)[1]) / (2 * population_size)
	f_B <- (2 * dim(BB)[1] + dim(AB)[1]) / (2 * population_size)
	
	cat(colnames(col), min(f_A, f_B), "\n", sep="\t")
}
```

To see how many alleles have MAF greater than 0.03, 0.05, and 0.1, we simply make some minor changes in the above `for` loop, and include a counter

```{r}
counter0 <- 0 # for checking > 0.1
counter1 <- 0 # for checking > 0.05
counter2 <- 0 # for checking > 0.03

for (i in 2:dim(dat)[2]) { # for each column
	col <- select(dat, i)
	
	AA <- filter(col, col==0)
	AB <- filter(col, col==1)
	BB <- filter(col, col==2)
	
	f_A <- (2 * dim(AA)[1] + dim(AB)[1]) / (2 * population_size)
	f_B <- (2 * dim(BB)[1] + dim(AB)[1]) / (2 * population_size)
	
	if (min(f_A, f_B) > 0.1)
		counter0 <- counter0 + 1
	if (min(f_A, f_B) > 0.05)
		counter1 <- counter1 + 1
	if (min(f_A, f_B) > 0.03)
		counter2 <- counter2 + 1
}

cat("# of SNPs with MAF greater than 0.03 is", counter2, "\n", sep=" ")
cat("# of SNPs with MAF greater than 0.05 is", counter1, "\n", sep=" ")
cat("# of SNPs with MAF greater than 0.1 is", counter0, "\n", sep=" ")
```

There are **8475** SNPs with MAF > 0.03, **7804** SNPs with MAF > 0.05, and **6495** SNPs with MAF > 0.1.


### Logistic Regression

$$ \theta(x) = Pr\{y=1 | x\} $$

$$ f(x) = \log\frac{\theta(x)}{1-\theta(x)} \sim \mu + \beta x $$

In our case, the outcome Y is binary. Let Y = 0 is for controls, Y = 1 is for cases. The predictors X are the genotypes we saw in the previous section: AA (X = 0), AB (X = 1), BB (X = 2), and missing (X = 3). However X = 3 is not present in our dataset.

We begin by loading and merging the *genotype.csv* and *phenotype.csv* files.

```{r}
library(dplyr) #  for manipulating tabular data

geno_dat <- read.csv("genotype.csv")
pheno_dat <- read.csv("phenotype.csv")
dat <- merge(pheno_dat, geno_dat)
```

For SNPs with MAF greater than at least 0.05, we perform logistic regression. Each SNP column is **factored** to indicate that it is to be used as a categorical variable (possible categories are 0, 1, 2).

```{r}
for (i in 2:dim(dat)[2]) { # for each column
	AA <- filter(dat, dat[,i]==0)
	AB <- filter(dat, dat[,i]==1)
	BB <- filter(dat, dat[,i]==2)
	f_A <- (2 * dim(AA)[1] + dim(AB)[1]) / (2 * population_size)
	f_B <- (2 * dim(BB)[1] + dim(AB)[1]) / (2 * population_size)

	# Two-way contigency table of outcome and predictors to make sure there are no empty cells
	#xtabs(~ V1 + dat[,i], data = dat)

	if (min(f_A, f_B) >= 0.05) {
		dat[,i] <- factor(dat[,i]) # to indicate that SNP column is to be used as a categorical variable
		logitfit <- glm(V1 ~ dat[,i], data = dat, family = binomial)
	}
}
```

Now, we can use the summary command.
```{r}
summary(logitfit)
```
```{r echo=false}

```
