---
title: "Finding cis-eQTLs"
output: pdf_document
layout: page
---

## Linear model for eQTL testing

```{r setup}
library(dplyr)

e <- read.csv("data/geuv_chr22_100genes_expression.csv", row.names=1) # Gene expressions
head(e)[,1:3] # Rows represent genes
start <- read.csv("data/geuv_chr22_100genes_position.csv", row.names=1) # Start site of genes
head(start)
snp <- read.csv("data/chr22_snp.csv", row.names=1) # SNP genotypes
head(snp)[,1:3]
snp.position <- read.csv("data/chr22_snp_position.csv", row.names=1) # SNP position
head(snp.position)
```

For each gene, we'll identify which SNPs are within 1000 bp of the gene start position (either upstream or downstream).

```{r cis}
eqtl <- list()
for (i in 1:nrow(start)) { # For each gene
  tmp <- vector()
  for (j in 1:nrow(snp.position)) { # For each SNP
    if (snp.position[j,2] >= start[i,2]-1000 & snp.position[j,2] <= start[i,2]+1000) {
      tmp <- append(tmp, rownames(snp.position)[j])
    }
  }
  eqtl[[i]] <- tmp
}

head(eqtl) # For each of the 100 genes, a list of SNPs +-1000 bp of the gene's start site
```

Now, for each SNP within 1000 bp of gene start position, we're going to check if the SNP is associated with expression of the gene using the linear regression model

$$ E_i = \mu + \beta_j G_{ij} + \epsilon_{ij} $$

where $E_i$ is expression of gene in individual $i$ and $G_{ij}$ is genotype of SNP $j$.

We're going to do this **for each gene**.

```{r regression}
# Setup gene expression matrix so that it's an appropriate input for lm()
E <- as.data.frame(t(e)) # Rows now represent individuals

obs_p <- vector()
all.M <- list()
for (i in 1:ncol(E)) { # For each gene
  if (length(eqtl[[i]]) != 0) {
    # Prepare genotype matrix for the given gene
    g <- snp[eqtl[[i]],]
    G <- as.data.frame(t(g))
    
    M <- lm(as.matrix(E[,i]) ~ as.matrix(G)) # Apply linear regression
    all.M[[i]] <- M # Save result for each gene
    obs_p <- append(obs_p, anova(M)$'Pr(>F)'[1])
  }
  else {
    all.M[[i]] <- NULL
    obs_p <- append(obs_p, 0)
  }
}
# anova(M)$'Pr(>F)'[1]
# summary(M)$coefficients[,4]
```

## Q-Q plot of p-values

```{r qqplot}
# Since these tests can give very small p-values, we transform by taking -log10(p-value)
obs_p = -log10(sort(obs_p))
# Creating theoretical p-values that follow a uniform distribution
th_p = -log10(seq(1/100,1,1/100))
plot(th_p,obs_p, xlab="Theoretical p-values", ylab="Observed p-values", main="Q-Q Plot")
lines(th_p,th_p,col='red')
```

## How many genes have at least one SNP with an association p-value less than 0.01?

```{r, echo=FALSE}
result.genes = vector()
result.snps = vector()
result.pvals = vector()

for (i in 1:ncol(E)) { # For each gene
  if (length(eqtl[[i]]) != 0) {
    # Prepare genotype matrix for the given gene
    g <- snp[eqtl[[i]],]
    G <- as.data.frame(t(g))
    
    M <- lm(as.matrix(E[,i]) ~ as.matrix(G)) # Apply linear regression
    
    len <- length(summary(M)$coefficients[,4])
    if (len > 1) {
      min.p <- min(as.vector(summary(M)$coefficients[2:len,4]))
      min.i <- which(as.vector(summary(M)$coefficients[2:len,4]) == min(as.vector(summary(M)$coefficients[2:len,4])))
    }
    if (min.p < 0.01) {
      result.genes <- append(result.genes, colnames(E)[i])
      result.snps <- append(result.snps, rownames(summary(M)$coefficients)[min.i+1])
      result.pvals <- append(result.pvals, min.p)
    }
  }
}

result.snps <- gsub("as\\.matrix\\(G\\)\\b", result.snps, replacement="") # Format SNP names
cbind(result.genes, result.snps, result.pvals)
```

